datafusion/core/src/physical_optimizer/replace_repartition_execs.rs

```diff
@@ -0,0 +1,781 @@
+/// The rule only works for conditions that have a `SortExec` and at least one `RepartitionExec`
+/// with multiple input partitioning.
+///
+/// The algorithm flow is simply like this:
+/// 1. Visit nodes of the physical plan top-down and look for a SortExec node (if not found do nothing)
+/// 2. If `SortExec` is found, iterate over its children recursively until the leaf node
+///      or another `SortExec` and check if anything breaks the ordering between
+///      `RepartitionExec`s with multiple input partitions considered as they maintain input ordering
+///       because they're potentially changed with `SortPreservingRepartitionExec`s
+/// 2_1. If the ordering is broken somehow, do nothing
+/// 2_2. If the ordering is kept the same among all children continue
+/// 3. Start a transform loop bottom-up on `SortExec`s children
+/// 3_1. Check if the child is a `RepartitionExec` with multiple input partitions and if the ordering matches with the `SortExec`
+/// 3_1_1. If conditions match, replace the `RepartitionExec` with `SortPreservingRepartitionExec`
+/// 3_1_2. If conditions do not match, keep the plan as is.
+/// 4. End the bottom-up transformation loop
+/// 5. Check if the `SortExec` plan is changed somehow.
+/// 5_1. If changed remove the `SortExec` from the plan
+/// 5_2. If not changed then keep the plan as is.
+/// 6. Continue to the top-down transformation
+pub struct ReplaceRepartitionExecs {}
+
+impl ReplaceRepartitionExecs {
+    #[allow(missing_docs)]
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+/// Creates a `SortPreservingRepartitionExec` from given `RepartitionExec`
+fn sort_preserving_repartition(
+    repartition: &RepartitionExec,
+) -> Result<Arc<SortPreservingRepartitionExec>> {
+    Ok(Arc::new(SortPreservingRepartitionExec::try_new(
+        repartition.input().clone(),
+        repartition.partitioning().clone(),
+    )?))
+}
+
+fn does_plan_maintains_input_order(plan: &Arc<dyn ExecutionPlan>) -> bool {
+    return plan.maintains_input_order().iter().all(|flag| *flag);
+}
+
+/// Check the children nodes between two `SortExec`s if they all maintain the input ordering
+/// if all of the related children maintain, then it is a possibility to replace `RepartitionExec`s
+fn replace_sort_children(
+    plan: &Arc<dyn ExecutionPlan>,
+) -> Result<Arc<dyn ExecutionPlan>> {
+    if plan.children().is_empty() {
+        return Ok(plan.clone());
+    }
+
+    let mut children = plan.children();
+    for (i, child) in enumerate(plan.children()) {
+        if !is_repartition(&child) && !does_plan_maintains_input_order(&child) {
+            break;
+        }
+
+        if let Some(repartition) = child.as_any().downcast_ref::<RepartitionExec>() {
+            if !repartition.maintains_input_order()[0] {
+                let spr = sort_preserving_repartition(repartition)?
+                    .with_new_children(repartition.children());
```

Comment:
```suggestion
                    .with_new_children(repartition.children())?;
```
doing error propagation here is better